shimmingtoolbox/b1/b1_shim.py

```diff
@@ -0,0 +1,270 @@
+
+        def cost(weights):
+            b1_abs = combine_maps(b1_roi, vector_to_complex(weights))
+            return np.square(np.linalg.norm(b1_abs - target)) + reg_param * (1 / np.mean(np.square(b1_abs)))
+
+    elif algo == 3:
+        # MLS targeting mean B1 at each iteration
+        def cost(weights):
+            b1_abs = combine_maps(b1_roi, vector_to_complex(weights))
+            return np.square(np.linalg.norm(b1_abs - np.mean(b1_abs))) + reg_param * (1 / np.mean(np.square(b1_abs)))
+
+    elif algo == 4:
+        # Maximizing the minimum B1+ value to get better RF efficiency
+        def cost(weights):
+            b1_abs = combine_maps(b1_roi, vector_to_complex(weights))
+            return 1 / np.min(b1_abs)
+
+    elif algo == 5:
+        # Minimizing the difference between the maximum and minimum B1+ values
+        def cost(weights):
+            b1_abs = combine_maps(b1_roi, vector_to_complex(weights))
+            # Using 3rd and 97th B1+ percentile values to minimize the influence of extreme values
+            return np.percentile(b1_abs, 97) - np.percentile(b1_abs, 3)
+
+    else:
+        raise ValueError(f"The specified algorithm does not exist. It must be an integer between 1 and 5.")
+
+    if constrained:
+        if sar_factor < 1:
+            raise ValueError(f"The SAR factor must be equal to or greater than 1.")
+        if q_matrix is None:
+            raise ValueError(f"A Q matrix must be provided in order to perform SAR constrained optimization.")
+        max_sar = sar_factor * sar(vector_to_complex(weights_init), q_matrix)
+        cons = ({'type': 'eq', 'fun': lambda w: np.linalg.norm(vector_to_complex(w)) - 1},  # Norm constraint
+                {'type': 'ineq', 'fun': lambda w: -sar(vector_to_complex(w), q_matrix) + max_sar})  # SAR constraint
+    else:
+        cons = ({'type': 'eq', 'fun': lambda w: np.linalg.norm(vector_to_complex(w)) - 1})  # Norm constraint only
+
+    shim_weights = vector_to_complex(scipy.optimize.minimize(cost, weights_init, constraints=cons).x)
+
+    return shim_weights
+
+
+def combine_maps(b1_maps, weights):
+    """
+    Combines the B1 field distribution of several channels into one map representing the total B1 field magnitude.
+
+    Args:
+        b1_maps (numpy.ndarray): Complex B1 field for different channels (x, y, n_slices, n_channels).
+        weights (numpy.ndarray): 1D complex array of length n_channels.
+
+    Returns:
+        numpy.ndarray: B1 field distribution obtained when applying the provided shim weights.
+    """
+    if b1_maps.shape[-1] == len(weights):
+        pass
+    else:
+        raise ValueError(f"The number of shim weights does not match the number of channels.\n"
+                         f"Number of shim weights: {len(weights)}\n"
+                         f"Number of channels: {b1_maps.shape[-1]}")
+
+    return np.abs(np.matmul(b1_maps, weights))
+
+
+def cov(array):
```

Comment:
probably something that can be used for b0 shimming as well

Reply:
scipy.stats.variation does something similar